src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorPatternInstance.java

```diff
@@ -0,0 +1,170 @@
+            patternInstances = new HashMap();
+
+        interfaceComponents.forEach((interfaceComponent) -> {
+            DecoratorPatternInstance patternInstance =
+                new DecoratorPatternInstance();
+            patternInstance.interfaceComponent = interfaceComponent;
+            patternInstances.putIfAbsent(interfaceComponent, patternInstance);
+        });
+
+        ArrayList<ClassOrInterfaceDeclaration> identifiedElements =
+            new ArrayList<>();
+
+        // Found no better way to populate PatternInstances than looping
+        // through everything multiple times
+        interfaceComponents.forEach((interfaceComponent) -> {
+            if (interfaceComponent.isInterface()) {
+                String interfaceName = interfaceComponent.getNameAsString();
+                concreteComponents.forEach(cc -> {
+                    cc.getImplementedTypes().forEach(implementedInterface -> {
+                        if (implementedInterface.getNameAsString().equals(
+                            interfaceName)) {
+                            DecoratorPatternInstance patternInstance =
+                                patternInstances.get(interfaceComponent);
+                            patternInstance.concreteComponents.add(cc);
+                            identifiedElements.add(cc);
+                        }
+                    });
+                });
+
+                abstractDecorators.forEach(ad -> {
+                    ad.getImplementedTypes().forEach(implementedInterface -> {
+                        if (implementedInterface.getNameAsString().equals(
+                            interfaceName)) {
+                            DecoratorPatternInstance patternInstance =
+                                patternInstances.get(interfaceComponent);
+                            patternInstance.abstractDecorators.add(ad);
+                            identifiedElements.add(ad);
+
+                            concreteDecorators.forEach(cd -> {
+                                cd.getExtendedTypes().forEach(extendedClass -> {
+                                    if (extendedClass.getNameAsString().equals(
+                                        ad.getName().asString())) {
+                                        patternInstance.concreteDecorators.add(
+                                            cd);
+                                        identifiedElements.add(cd);
+                                    }
+                                });
+                            });
+                        }
+                    });
+                });
+                // If any elements are left then they are invalid instances
+                concreteComponents.removeAll(identifiedElements);
+                abstractDecorators.removeAll(identifiedElements);
+                concreteDecorators.removeAll(identifiedElements);
+            } else {
+                throw new UnsupportedOperationException("Was not an interface");
```

Comment:
The only case that this else will be called is if list contains Classes as well as Interfaces, correct? If so, then the class has been marked incorrectly and it should be reported with a feedback.

Reply:
Or wait, is that handled below?